File: java/linked_list.py


class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def add_node(self,data):
        if self.head == None:
            self.head = Node(data)
            return 
        temp = self.head
        while(temp.next):
            temp = temp.next
        temp.next = Node(data)
    
    def insert_at_beginning(self,data):
        node = Node(data)
        node.next = self.head
        self.head = node
    
    def insert_values(self,data_list):
        for i in data_list:
            self.add_node(i)

    # def remove(self, index):
    #     if index < 0 or index > self.length():
    #         raise Exception("Invalid Index")
    #     else:
    #         if index == 0:
    #             self.head = self.head.next

    #     count = 0
    #     itr = self.head
    #     while itr:
    #         if count == index - 1:
    #             itr.next = itr.next.next
    #             break

    #         itr = itr.next
    #         count+=1
    def remove(self,val):
        if self.head.data == val:
            self.head = self.head.next

        temp = self.head.next
        prev = self.head
        while (temp.next):
            if temp.data == val:
                temp = temp.next
                prev.next = temp

            prev = temp
            temp = temp.next
        
        prev.next = temp.next
        
        


        

    
    def print(self):
        temp = self.head
        while(temp):
            if temp.next:
                print(temp.data, end = '-->')
            else:
                print(temp.data)
            temp = temp.next

    
    def length(self):
        if self.head == None:
            return 0
        else:
            temp = self.head
            count = 1
            while (temp.next):
                count += 1
                temp = temp.next
            return count 
        
        
if __name__ == '__main__':
    llist = LinkedList()
    llist.add_node(2)
    llist.add_node(3)
    llist.insert_at_beginning(1)
    #llist.insert_values(['jdjk','dhfk','hdsffh'])
    #llist.remove(1)
    llist.remove(3)
    llist.print()

    print('length of linked list',llist.length())

File: leetcode/LinkedList.java
public class LinkedList {
	

	static class Node {
		int data;
		Node next;

	}
	Node head;
	
	public void insert(int d){
		Node node = new Node();
		node.data = d;
		node.next = null;

		if (head == null){
			head = node;
		}
		else{
			Node n = new Node();
			n = head;
			while (n.next != null){
				n = n.next;
			}
			n.next = node;
		}
	}

	public void insert_start(int d){
		Node node = new Node();
		node.data = d;
		node.next = head;
		head = node;
		
	}
	public void del_position(int position){
		
		Node temp = head;
		if (position == 0){
			head = temp.next;
		}
		//get the prev node of the current position
		for (int i = 0; temp != null & i < position -1; i++) {
			temp = temp.next;
		}
		if (temp == null || temp.next == null)
            return;

		temp.next = temp.next.next;


	}
	public void delete(int key){
		Node temp = head;
		Node prev = null;
		//if head node is the node to del
		if (temp != null &&  temp.data == key){
			head = temp.next;
			return;
		}
		//middle of the linked list
		while(temp != null && temp.data != key){
			prev = temp;
			temp = temp.next;
		}
		//not in linked list
		if (temp == null)
			return;
		//if key value finds 
		prev.next = temp.next;

	}
	public  int len(){
		int l = 0;
		Node temp = head;
		while (temp.next != null){
			l += 1;
			temp = temp.next;
		}
		return l;
	}

	public void print()
	{
		Node n = new Node();
		n = head;
		while(n.next != null){
			System.out.println(n.data);
			n = n.next;
		}
		System.out.println(n.data);
	}
	
	public static void main(String[] args) {
		LinkedList list = new LinkedList();
		list.insert(1);
		list.insert(2);
		list.insert_start(0);
		list.insert_start(9);
		//list.del_position(2);
		list.print();
		System.out.println(list.len());
	}


}


